### ISSUE TITLE
How long did you pertained the OSCAR/OSCAR+

### ISSUE THREAD

**Issue by FingerRec** [microsoft/Oscar#110]
The GPU resources and training time?

**Comment by coldmanck** [comment]
@pzzhang Really appreciate it if you could clarify this!

**Comment by pzzhang** [comment]
Hi @FingerRec  and @coldmanck , please change the batch size to 1024, which is the batch size reported in the paper.

Sorry that the batch size 8 here is just for debugging.

**Comment by coldmanck** [comment]
@pzzhang Thank you for your reply! How about the training resources (gpu type/memory/time) you used? 

I'm working on less capable GPUs, so I am also wondering how I should modify the batch size accordingly (and its affect to the performance). Please let me know if you have any idea/advice on this! Thank you a lot :)

**Comment by coldmanck** [comment]
Hi @pzzhang I see that I need ~22 days to pretrain OSCAR+ on 8 V100 GPUs, each 21GB memory occupied. As V100 has 32G memory available, I am considering increase the batch size to 1.5X larger (i.e., 1536) and maybe I'd only need 2/3 of the required steps. Do you think it's valid? 

Moreover, I'd like you to clarify what do you mean by **OSCAR+ is trained at least 1M steps**. I see that `max_iter` is set as 2M by default, and I couldn't find code about early stopping here; so I am wondering what does **at least** mean here. Also, the 2M `max_iter` (which affects the speed of linear learning rate decay). If we want to train, say, (at least?) 1M steps, shouldn't we modify it to 1M accordingly?

**Comment by pzzhang** [comment]
@coldmanck  
training resources:  16 V100 (32G), 1024 total batchsize, train for 2M iterations takes about 20 weeks. We do not use AMP, and the 32G memory is sufficient. If you use AMP, the memory and training time can be reduced.

We just evaluate the checkpoint in the middle, for every 100k iterations. We pick the best checkpoint among these evaluated checkpoints. 

You do not need to modify it to 1M accordingly, because the learning rate is linear and it will be too small for the final few iterations.

**Comment by coldmanck** [comment]
@pzzhang Thank you for your detailed reply. I really appreciate it! :) 

I have another question: VinVL consists of two types of training samples, namely, _<caption-tags-img_features>_ or _<question-answer-img_features>_. It seems natural to me that the examples of the former type should be used to predict polluted **captions**, while the latter should be used to predict polluted **answers**. However, it seems that [your code](https://github.com/microsoft/Oscar/blob/df79152b708c3c46f2dc93324776a27406ccc634/oscar/datasets/oscar_tsv.py#L256) does not distinguish the type of input examples. Chances are a QA pair gets question replaced, and a caption-tags pair gets tags replaced. Have I missed anything?

**Comment by ckmstydy** [comment]
@coldmanck Hello, When you did the pre-training, small corpus and large carpus take the same amount of time? When I tried，why is it so much slower to pre-train with a large corpus than with a small corpus?

**Comment by VivaLaDijkstra** [comment]
> Hi @pzzhang I see that I need ~22 days to pretrain OSCAR+ on 8 V100 GPUs, each 21GB memory occupied. As V100 has 32G memory available, I am considering increase the batch size to 1.5X larger (i.e., 1536) and maybe I'd only need 2/3 of the required steps. Do you think it's valid?
> 
> Moreover, I'd like you to clarify what do you mean by **OSCAR+ is trained at least 1M steps**. I see that `max_iter` is set as 2M by default, and I couldn't find code about early stopping here; so I am wondering what does **at least** mean here. How did you choose the final checkpoint to report the performance?
> 
> Also, the 2M `max_iter` (which affects the speed of linear learning rate decay). If we want to train, say, (at least?) 1M steps, shouldn't we modify it to 1M accordingly?

20 weeks? 140 days? Seriously? Or what you really want to say is 20 days?

**Comment by ignasa007** [comment]
> > Hi @pzzhang I see that I need ~22 days to pretrain OSCAR+ on 8 V100 GPUs, each 21GB memory occupied. As V100 has 32G memory available, I am considering increase the batch size to 1.5X larger (i.e., 1536) and maybe I'd only need 2/3 of the required steps. Do you think it's valid?
> > Moreover, I'd like you to clarify what do you mean by **OSCAR+ is trained at least 1M steps**. I see that `max_iter` is set as 2M by default, and I couldn't find code about early stopping here; so I am wondering what does **at least** mean here. How did you choose the final checkpoint to report the performance?
> > Also, the 2M `max_iter` (which affects the speed of linear learning rate decay). If we want to train, say, (at least?) 1M steps, shouldn't we modify it to 1M accordingly?
> 
> 20 weeks? 140 days? Seriously? Or what you really want to say is 20 days?

Not sure if it'd be 20 days. 20 days would mean 20*86_400/2_000_000 = 0.864 seconds per iteration, but time for end-to-end inference itself is over half a second (last page of the paper), and I think it is only for one sample. Add to that loading the data for a whole batch, computing gradients, updating parameters, logging, etc. hard to believe it'd take 0.864 seconds. Rather 20 weeks seems more reasonable to me, which comes out to be 0.864*7 ≈ 6 seconds per batch. But regardless, an insanely long time to pre-train a model; blew my mind!

PS. I'm an amateur, so might be wrong with my interpretations :')
